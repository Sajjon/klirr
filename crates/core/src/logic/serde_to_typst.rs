use crate::prelude::*;
use indoc::indoc;

/// Empty Marker trait
pub trait ToTypst {}

pub trait ToTypstFn: ToTypst {
    /// Converts the implementing type into a Typst function returning a dictionary.
    fn to_typst_fn(&self) -> String;
}

impl<T: ToTypst + Serialize> ToTypstFn for T {
    /// Converts this  `Serialize`able Rust struct into Typst syntax.
    fn to_typst_fn(&self) -> String {
        let value = serde_json::to_value(self).expect("Serialization failed");
        format!(
            indoc! {r#"
        #let provide() = {{
          {}
        }}
        "#},
            to_typst_value(&value, 0)
        )
    }
}

/// Recursively converts a serde_json::Value into pretty-printed Typst syntax.
fn to_typst_value(value: &Value, indent: usize) -> String {
    let indent_str = "  ".repeat(indent);
    let next_indent = indent + 1;
    let next_indent_str = "  ".repeat(next_indent);

    match value {
        Value::Object(map) => {
            // Flatten single-entry enum-like objects (e.g. { "Net": 30 }) to (net: 30)
            if map.len() == 1 {
                if let Some((variant, inner)) = map.iter().next() {
                    if inner.is_number() || inner.is_string() || inner.is_object() {
                        return format!(
                            "(\n{}{}: {},\n{})",
                            next_indent_str,
                            variant.to_lowercase(),
                            to_typst_value(inner, next_indent),
                            indent_str
                        );
                    }
                }
            }

            let fields = map
                .iter()
                .map(|(k, v)| {
                    format!(
                        "{}{}: {}",
                        next_indent_str,
                        k,
                        to_typst_value(v, next_indent)
                    )
                })
                .collect::<Vec<_>>()
                .join(",\n");

            format!("(\n{},\n{})", fields, indent_str)
        }

        Value::Array(arr) => {
            let items = arr
                .iter()
                .map(|v| format!("{}{}", next_indent_str, to_typst_value(v, next_indent)))
                .collect::<Vec<_>>()
                .join(",\n");

            format!("(\n{},\n{})", items, indent_str)
        }

        Value::String(s) => format!("\"{}\"", s),
        Value::Bool(b) => b.to_string(),
        Value::Number(n) => n.to_string(),
        Value::Null => "none".to_string(),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use test_log::test;

    macro_rules! test_data_to_typst {
        ($sample:expr, $input:expr) => {{
            let input = $sample
                .to_partial($input)
                .unwrap()
                .to_typst(
                    ExchangeRates::builder()
                        .target_currency(Currency::EUR)
                        .rates(ExchangeRatesMap::from_iter([
                            (Currency::GBP, UnitPrice::from(dec!(1.174))),
                            (Currency::SEK, UnitPrice::from(dec!(11.05))),
                        ]))
                        .build(),
                )
                .unwrap();
            let typst = input.to_typst_fn();
            insta::assert_snapshot!(typst);
        }};
    }

    macro_rules! test_l18n_to_typst {
        ($input:expr) => {{
            let typst = $input.content().to_typst_fn();
            insta::assert_snapshot!(typst);
        }};
    }

    #[test]
    fn sample_expenses_to_typst() {
        test_data_to_typst!(
            Data::sample(),
            ValidInput::builder()
                .items(InvoicedItems::Expenses)
                .period(YearAndMonth::sample())
                .language(Language::EN)
                .build()
        );
    }

    #[test]
    fn sample_services_to_typst() {
        test_data_to_typst!(
            Data::sample(),
            ValidInput::builder()
                .items(InvoicedItems::Service { time_off: None })
                .period(YearAndMonth::sample())
                .language(Language::EN)
                .build()
        );
    }

    #[test]
    fn l18n_english_to_typst_macro() {
        test_l18n_to_typst!(&L18n::new(Language::EN).unwrap());
    }
}
