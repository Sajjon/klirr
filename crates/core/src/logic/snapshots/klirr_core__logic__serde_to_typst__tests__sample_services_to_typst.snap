---
source: crates/core/src/logic/serde_to_typst.rs
expression: typst
---
#let provide() = {
  (
  client: (
    company_name: "Holmes Ltd",
    contact_person: "Sherlock Holmes",
    organisation_number: "9876543-2101",
    postal_address: (
      city: "London",
      country: "England",
      street_address: (
        line_1: "221B Baker Street",
        line_2: "",
      ),
      zip: "NW1 6XE",
    ),
    vat_number: "GB987654321",
  ),
  information: (
    due_date: "2025-06-30",
    emphasize_color_hex: "#8b008b",
    footer_text: "Billed with the utmost discretion—your secrets are safe, for a price.",
    invoice_date: "2025-05-31",
    number: 33,
    purchase_order: "PO-12345",
  ),
  line_items: (
    is_expenses: false,
    items: (
      (
        currency: "EUR",
        name: "Discreet Investigative Services",
        quantity: 23.0,
        total_cost: 17871.0,
        transaction_date: "2025-05-31",
        unit_price: 777.0,
      ),
    ),
  ),
  output_path: (
    name: "2025-05-31_Lupin_et_Associés_invoice_33.pdf",
  ),
  payment_info: (
    bank_name: "Banque de Paris",
    bic: "BNPAFRPP",
    currency: "EUR",
    iban: "FR76 3000 6000 0112 3456 7890 189",
    terms: "Net 30",
  ),
  vendor: (
    company_name: "Lupin et Associés",
    contact_person: "Arsène Lupin",
    organisation_number: "7418529-3012",
    postal_address: (
      city: "Paris",
      country: "France",
      street_address: (
        line_1: "5 Avenue Henri-Martin",
        line_2: "Appartement 24",
      ),
      zip: "75116",
    ),
    vat_number: "FR74185293012",
  ),
)
}
